### PR TITLE
First pass review for Android file storage

### PR DESCRIPTION
**Pull request by commjoen**
Even though this branch is focussed on iOS, I had to update the android first. Hope to do 2 PRs from the branch, but I wanted to get started by just updating the Android part so I can mimic in in iOS as well.

### CODE REVIEW

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 1]
```diff
@@ -675,17 +677,126 @@ private void crcTest() throws IOException {
  }
 }
 ```
+##### Sample Implementation - Storage
+
+When providing integrity on the file-storage itself. You can either create an HMAC over a given key-value pair as for the Android `SharedPreferences` or you can create an HMAC over a complete file provided by the filesystem.
+When using an HMAC, you can either use a bouncy castle implementation to HMAC the given content or the AndroidKeyStore and then verify the HMAC later on. There are a few steps to take care off.
```
Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Extra "f" at the end of "of"

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 2]
```diff
@@ -675,17 +677,126 @@ private void crcTest() throws IOException {
...
+When verifying the HMAC with BC:
+
+1. Make sure BounceyCastle or SpongeyCastle are registered as a security provider.
+2. Extract the message and the hmacbytes as separate arrays.
+3. repeate step 1-4 of generating an hmac on the message.
+4. now compare the extracted hamcbytes to the result of step 3.
+
+When generating the HMAC based on the Android keystore, then it is best to only do this for Android 6 and higher. In that case you generate the key for hmacking as described in [3].
+
+----TODO: ADD REFERENCE TO SAMPLE IMPELEMTNATION CLASSES, o ruse:
```
Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: probably meant to say "or use"

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 3]
```diff
@@ -675,17 +677,126 @@ private void crcTest() throws IOException {
...
+
+```java
+public enum HMACWrapper {
+    HMAC_512("HMac-SHA512"),
```
Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Isn't this supposed to be "HmacSHA512"?

Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Nvm, it appears that both variants work: capitalization and presence/absence of the dash don't seem to matter, at least on the Nougat emulator

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md (hunk 2 above)
Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Also, check the spelling of Implementation

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 4]
```diff
@@ -675,17 +677,125 @@ private void crcTest() throws IOException {
...
+
 1. Patch out the anti-debugging functionality. Disable the unwanted behaviour by simply overwriting the respective bytecode or native code it with NOP instructions.
 2. Use Frida or Xposed to hook APIs to hook file system APIs on the Java and native layers. Return a handle to the original file instead of the modified file.
 3. Use Kernel module to intercept file-related system calls. When the process attempts to open the modified file, return a file descriptor for the unmodified version of the file instead.
 
 Refer to the "Tampering and Reverse Engineering section" for examples of patching, code injection and kernel modules.
 
+*When trying to bypass the storage integrity checks*
+
+1. retrieve the data from the device, as described at the secion for device binding.
```
Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Capitalize "Retrieve"

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 5]
```diff
@@ -675,17 +677,125 @@ private void crcTest() throws IOException {
...
+2. Extract the message and the hmacbytes as separate arrays.
+3. repeate step 1-4 of generating an hmac on the message.
+4. now compare the extracted hamcbytes to the result of step 3.
+
+When generating the HMAC based on the Android keystore, then it is best to only do this for Android 6 and higher. In that case you generate the key for hmacking as described in [3].
+A convinient HMAC implementation without the `AndroidKeyStore` can be found below:
+
+```java
+public enum HMACWrapper {
+    HMAC_512("HMac-SHA512"),
```
Comment (bolot) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: FYI: turns out capitalization and presence/absence of the dash ("-") does not matter, at least on the Nougat emulator.

Comment (commjoen) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Added comment to clear this

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md (hunk 4 above)
Comment (commjoen) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Fixed! thanks!

**Pull request merged**